[utils] revamp options controlling lit's output

hnrklssn opened the pull request:
Lit has a number of options controlling the output, but they don't compose very well. This breaks the existing options down into smaller, orthogonal options, and makes the existing options aliases of the new ones.

This introduces the following options:
  --test-output {off,failed,all}
  --print-result-after {off,failed,all}
  --diagnostic-level {error,warning,note}
  --terse-summary
  --no-terse-summary
  --progress-bar (mirroring --no-progress-bar)

--test-output and --print-result-after are not entirely orthogonal, as '--test-output X' requires that --print-result-after is set to at least X, and implicitly does so if it isn't already. Conversely, '--print-result-after Y' requires that --test-output is at most Y, and implicitly lowers if it is higher. This means that the following invocations have different end results, as they are applied in order:
  '--test-output all --print-result-after off'
  '--print-result-after off --test-output all'

The following existing options are now aliases as follows:
  -q, --quiet
    '--diagnostic-level error --test-output off --terse-summary'
  -s, --succinct
    '--progress-bar --print-result-after failed'
  -v, --verbose
    '--test-output failed'
  -a, --show-all
    '--test-output all'

These where all completely separate options and would override each other in ad-hoc ways, with no regard to the order they were given.

This fixes https://github.com/llvm/llvm-project/issues/106643 This is based on the RFC
https://discourse.llvm.org/t/rfc-new-command-line-options-for-controlling-llvm-lit-output/ with the addition of --terse-summary, which was a behaviour of -q that was not captured by the original RFC. This also diverges from the RFC in that --debug is NOT folded into --diagnostic-level, because it can be useful to debug any configuration, including those specifying --diagnostic-level.

Example combination that is possible now but wasn't before: '--diagnostic-level error --test-output all --progress-bar' Another use case is aliases, where you can alias e.g:
  alias lit=llvm-lit --quiet
but still override the specified default options.

hnrklssn committed:
[utils] revamp options controlling lit's output

Lit has a number of options controlling the output, but they don't
compose very well. This breaks the existing options down into smaller,
orthogonal options, and makes the existing options aliases of the new
ones.

This introduces the following options:
  --test-output {off,failed,all}
  --print-result-after {off,failed,all}
  --diagnostic-level {error,warning,note}
  --terse-summary
  --no-terse-summary
  --progress-bar (mirroring --no-progress-bar)

--test-output and --print-result-after are not entirely orthogonal, as
'--test-output X' requires that --print-result-after is set to at least
X, and implicitly does so if it isn't already. Conversely,
'--print-result-after Y' requires that --test-output is at most Y, and
implicitly lowers if it is higher. This means that the following
invocations have different end results, as they are applied in order:
  '--test-output all --print-result-after off'
  '--print-result-after off --test-output all'

The following existing options are now aliases as follows:
  -q, --quiet
    '--diagnostic-level=error --test-output=off --terse-summary'
  -s, --succinct
    '--progress-bar --test-progress=failed'
  -v, --verbose
    '--test-output=failed'
  -a, --show-all
    '--test-output=all'

These where all completely separate options and would override each
other in ad-hoc ways, with no regard to the order they were given.

This fixes https://github.com/llvm/llvm-project/issues/106643
This is based on the RFC
https://discourse.llvm.org/t/rfc-new-command-line-options-for-controlling-llvm-lit-output/
with the addition of --terse-summary, which was a behaviour of -q that
was not captured by the original RFC. This also diverges from the RFC in
that --debug is NOT folded into --diagnostic-level, because it can be
useful to debug any configuration, including those specifying
--diagnostic-level.

Example combination that is possible now but wasn't before:
'--diagnostic-level error --test-output all --progress-bar'
Another use case is aliases, where you can alias e.g:
  alias lit=llvm-lit --quiet
but still override the specified default options.

hnrklssn committed:
format python, remove debug prints

hnrklssn committed:
remove one last debug print

hnrklssn committed:
update call to LitConfig

hnrklssn committed:
update call to LitConfig

hnrklssn committed:
[utils] fix lit -vv alias, remove rendundant uses of '-a -v'

-vv should be the same as -v

A bunch of lit tests use both -a and -v (or -vv), which was always
redundant, but now results in the last option being applied, rather than
-a always overriding -v.

ldionne reviewed (approved):
I am a big fan of your RFC and this PR. I've always found Lit's verbosity options to be confusing, and IMO this makes them a lot easier to work with. I'd almost be in favour of dropping the old options, but you're hugely decreasing the burden of this PR by keeping backwards compatibility, so I think that's the smart thing to do.

I have some comments, but this basically LGTM. I pulled your PR and ran the libc++ test suite with your patch applied, and my usual workflow was working as expected.

You probably want to have other people's feedback as well, but from my side this is 👍🏻.

hnrklssn commented on llvm/utils/lit/tests/Inputs/verbosity/lit.cfg:
```diff
@@ -0,0 +1,12 @@
+lit_config.dbg("this is a debug log")
+lit_config.note("this is a note")
+lit_config.warning("this is a warning")
+#lit_config.error("this is an error")
```
Ah no, good find!

hnrklssn committed:
remove redundant `if lit_config.debug`, and add help test detail

Checking lit_config.debug is redundant now, since calling lit_config.dbg
does the same check behind the scenes. lit_config.debug is kept around
for now to avoid breaking downstream users.

hnrklssn commented:
> @hnrklssn I think there might be a race condition in the [verbosity.py](http://verbosity.py) test you added:
> 
> https://lab.llvm.org/buildbot/#/builders/46/builds/26284 https://lab.llvm.org/buildbot/#/builders/190/builds/30855

Thanks for letting me know. This should fix it: https://github.com/llvm/llvm-project/pull/167801
